pmlc/conversion/gpu/gpu_to_vulkan.cc

```diff
@@ -237,10 +262,86 @@ Value ConvertGpuLaunchFuncToVulkanCalls::createEntryPointNameConstant(
                                   shaderName, LLVM::Linkage::Internal);
 }
 
+llvm::SmallSet<Operation *, 4>
+ConvertGpuLaunchFuncToVulkanCalls::getExternalDependentOperations(
+    Value &value) {
+  llvm::SmallSet<Operation *, 4> operations;
+  llvm::SetVector<OpOperand *> uses;
+
+  operations.insert(value.getDefiningOp());
+  for (auto &use : value.getUses()) {
+    uses.insert(&use);
+  }
+  while (!uses.empty()) {
+    auto owner = uses.back()->getOwner();
+    operations.insert(owner);
+    uses.pop_back();
+    for (auto opResult : owner->getOpResults()) {
+      for (auto &use : opResult.getUses()) {
+        uses.insert(&use);
+      }
+    }
+  }
+
+  for (auto operation : operations) {
+    if (isInternalOperation(operation)) {
+      operations.erase(operation);
+    }
+  }
+  return operations;
+}
+
+bool ConvertGpuLaunchFuncToVulkanCalls::isInternalOperation(Operation *op) {
+  // Buffer related calls are all mlir::CallOp, does not need to check
+  // LLVM::CallOp
+  auto callOp = llvm::dyn_cast<mlir::CallOp>(op);
+  if (callOp && (callOp.callee() == kBindAllBuffers ||
+                 optionalSymbols.count(callOp.callee()))) {
+    return true;
+  }
+
+  auto allocOp = llvm::dyn_cast<AllocOp>(op);
+  auto memRefCastOp = llvm::dyn_cast<MemRefCastOp>(op);
+  if (allocOp || memRefCastOp) {
+    return true;
+  }
+  return false;
+}
+
+uint32_t
+ConvertGpuLaunchFuncToVulkanCalls::getBufferCopyMode(mlir::CallOp &callOp,
+                                                     Value &buffer) {
+  uint32_t copyMode = kBufferCopyModeInit;
+  if (buffer.isa<BlockArgument>()) {
+    copyMode |= (kBufferCopyModeHostToDevice | kBufferCopyModeDeviceToHost);
+    return copyMode;
+  }
+
+  auto operationDeps = getExternalDependentOperations(buffer);
+  auto currentBlock = callOp.getOperation()->getBlock();
+  for (auto op : operationDeps) {
+    if (op->getBlock() == currentBlock) {
+      if (op->isBeforeInBlock(callOp.getOperation())) {
+        copyMode |= kBufferCopyModeHostToDevice;
+      } else if (deinitVulkan.getOperation()->isBeforeInBlock(op)) {
```

Comment:
I had some trouble reading this code given that the state we are conditioning on changes from line 324 (if dependency is before the *call op*) and line 326 (if *de initialization* is before the dependency).

I think it would read quite a lot more readable if you did this:
```
if dependency is before the call op then 
  set the copy-host-to-device bit
else // dependency is after the call op
  if de initialization is before the dependency then
    set the copy device-to-host bit
  else
    warning
```

I am wondering about this warning, though.  Under what circumstances would we hit this case?  Does it have to do with whether or not we are printing the buffers?  And, if it is just a warning is it worth checking?  Goes back to my design question from earlier --- maybe makes sense just to drop this check.  In which case you have...

```
if dependency is before the call op then 
  set the copy-host-to-device bit
else // dependency is after the call op
  set the copy device-to-host bit
```

And, no need to track the deinit across two member functions.

Reply:
Because currently we are doing calculations on both host and device side, thus some backend tests run into a situation where the host buffers are used after submit to vulkan and before returns, such as in testStd,
 
Below is a section of the IR for testStd
```
 func @get_value(%arg0: memref<3x4xf32> {tile.name = "anon"}, %arg1: memref<f32>) {
    %cst = constant 0.000000e+00 : f32
    %c12_i32 = constant 12 : i32
    %c0 = constant 0 : index
    %c3 = constant 3 : index
    %c4 = constant 4 : index
    %c1 = constant 1 : index
    %0 = alloc() : memref<1x1xf32>
    store %cst, %0[%c0, %c0] : memref<1x1xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %arg0, %0) {kernel = @get_value_kernel::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<1x1xf32>) -> ()
    %1 = alloc() : memref<1x1xf32>
    %2 = load %0[%c0, %c0] : memref<1x1xf32>
    %3 = sitofp %c12_i32 : i32 to f32
    %4 = divf %2, %3 : f32
    store %4, %1[%c0, %c0] : memref<1x1xf32>
    %5 = alloc() : memref<3x4xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c3, %c4, %arg0, %1, %5) {kernel = @get_value_kernel_0::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<1x1xf32>, memref<3x4xf32>) -> ()
    %6 = alloc() : memref<3x4xf32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c3, %c4, %5, %6) {kernel = @get_value_kernel_1::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<3x4xf32>) -> ()
    %7 = alloc() : memref<f32>
    store %cst, %7[] : memref<f32>
    "gpu.launch_func"(%c1, %c1, %c1, %c1, %c1, %c1, %6, %7) {kernel = @get_value_kernel_2::@get_value_kernel} : (index, index, index, index, index, index, memref<3x4xf32>, memref<f32>) -> ()
    %8 = alloc() : memref<f32>
    %9 = load %7[] : memref<f32>
    %10 = divf %9, %3 : f32
    store %10, %8[] : memref<f32>
    %11 = load %8[] : memref<f32>
    %12 = sqrt %11 : f32
    store %12, %arg1[] : memref<f32>
    return
  }
```
the code section below between the first and second gpu.launch_func is changing host side buffers after submit to vulkan, so it will end up with an incorrect result. But if the operation is stx.reshape, the result may still be correct.
```
    %1 = alloc() : memref<1x1xf32>
    %2 = load %0[%c0, %c0] : memref<1x1xf32>
    %3 = sitofp %c12_i32 : i32 to f32
    %4 = divf %2, %3 : f32
    store %4, %1[%c0, %c0] : memref<1x1xf32>
    %5 = alloc() : memref<3x4xf32>
```

I have detected a list of backend tests that currently has this issue, as follows,
FAIL: testBranchWithSoftmax (__main__.TestBackendOps) . 
FAIL: testRandomNormalDev (__main__.TestBackendOps) 
FAIL: testRandomUniformDev (__main__.TestBackendOps)  
FAIL: testRandomUniformVariableDev (__main__.TestBackendOps)  
FAIL: testStd (__main__.TestBackendOps) 
FAIL: testTruncatedNormalDev (__main__.TestBackendOps)  
FAIL: testVarSimple (__main__.TestBackendOps)  

Do you think put this message in IVLOG is better than warning?

To solve this issue we can 
1. Put everything between gpu.launch_func into a gpu module so they will be send to device
2. Rearrange the algorithm/pipeline passes to avoid changing host side buffers after copied to device
3. Make several vulkan command buffers and run the gpu.launch_funcs(and copy buffers back) before these host side ops.

Reply:
Addressing the larger design / testing issue in follow-on PRs works for me.  I see that you rewrote and commented the conditional logic in `getBufferCopyMode` and that looks good, thanks.